### PR TITLE
devicetree: Add clock-names support

### PR DESCRIPTION
**Pull request by galak**

### CODE REVIEW

File: include/devicetree/clocks.h [hunk 1]
```diff
@@ -54,6 +54,39 @@ extern "C" {
...
 
+/**
+ * @brief Get clock controller "name" (label property) by name
+ *
+ * It's an error if the clock controller referenced by the phandle
+ * in property "clocks" at index "idx" has no label property.
+ *
+ * Example devicetree fragment:
+ *
+ *     clk1: clkctrl@... {
```
Comment (mbolivar-nordic) [inline include/devicetree/clocks.h:65]: I took a look, and it seems like dtspec recommends either 'clock' or 'clock-controller' for these. Think it's worth changing the examples to align with the spec as follow up work? I know this is just using the existing convention so it's fine for me to keep it as is for this PR.

File: include/devicetree/clocks.h [hunk 2]
```diff
@@ -54,6 +54,39 @@ extern "C" {
...
+ *
+ *     DT_CLOCKS_LABEL_BY_NAME(DT_NODELABEL(n), beta) // "CLK_2"
+ *
+ * @param node_id node identifier
+ * @param name lowercase-and-underscores "clock" name
+ * @return the label property for the referenced node by name
+ * @see DT_PHANDLE_BY_NAME()
+ */
+#define DT_CLOCKS_LABEL_BY_NAME(node_id, name) \
+	DT_PROP(DT_PHANDLE_BY_NAME(node_id, clocks, name), label)
```
Comment (mbolivar-nordic) [inline include/devicetree/clocks.h:88]: You can just use DT_LABEL(DT_PHANDLE_BY_NAME(...)) if you prefer.

File: include/devicetree/clocks.h [hunk 3]
```diff
@@ -96,6 +129,41 @@ extern "C" {
 #define DT_CLOCKS_CELL_BY_IDX(node_id, cell, idx) \
 	DT_PHA_BY_IDX(node_id, clocks, idx, cell)
 
+/**
+ * @brief Get Clock controller "cell" value by name
```
Comment (mbolivar-nordic) [inline include/devicetree/clocks.h:133]:
```suggestion
 * @brief Get clock controller "cell" value by name
```

File: include/devicetree/clocks.h [hunk 4]
```diff
@@ -96,6 +129,41 @@ extern "C" {
...
+ *       - bus
+ *       - bits
+ *
+ * Example usage:
+ *
+ *     DT_CLOCKS_CELL_BY_NAME(DT_NODELABEL(n), alpha, bus) // 10
+ *     DT_CLOCKS_CELL_BY_NAME(DT_NODELABEL(n), beta, bits) // 40
+ *
+ * @param node_id node identifier
+ * @param cell binding's cell name within the specifier referenced as "name"
```
Comment (mbolivar-nordic) [inline include/devicetree/clocks.h:159]: Doxygen has cell, then name, but the macro has name, then cell, as the parameter order.

File: include/devicetree/clocks.h [hunk 5]
```diff
@@ -96,6 +129,41 @@ extern "C" {
...
+ * @param name lowercase-and-underscores specifier name
+ * @return the value of the cell inside the named specifier by name
```
Comment (mbolivar-nordic) [inline include/devicetree/clocks.h:161]:
```suggestion
 * @return the value of the cell inside the named specifier
```

Just a bit of redundancy there

File: include/devicetree/clocks.h [hunk 6]
```diff
@@ -115,6 +183,16 @@ extern "C" {
 #define DT_INST_CLOCKS_LABEL_BY_IDX(inst, idx) \
 	DT_CLOCKS_LABEL_BY_IDX(DT_DRV_INST(inst), idx)
 
+/**
+ * @brief Get a DT_DRV_COMPAT clock controller "name" by name
```
Comment (mbolivar-nordic) [inline include/devicetree/clocks.h:187]:
```suggestion
 * @brief Get a DT_DRV_COMPAT clock controller "name" (label property) by name
```

File: include/devicetree/clocks.h [hunk 7]
```diff
@@ -133,6 +211,16 @@ extern "C" {
 #define DT_INST_CLOCKS_CELL_BY_IDX(inst, cell, idx) \
 	DT_CLOCKS_CELL_BY_IDX(DT_DRV_INST(inst), cell, idx)
 
+/**
+ * @brief Get a DT_DRV_COMPAT clock controller "cell" value by name
+ * @param inst instance number
```
Comment (mbolivar-nordic) [inline include/devicetree/clocks.h:216]:
```suggestion
 * (see @ref DT_CLOCKS_CELL_BY_NAME)
 * @param inst instance number
```

File: include/devicetree/clocks.h [hunk 8]
```diff
@@ -133,6 +211,16 @@ extern "C" {
...
+ * @param name lowercase-and-underscores specifier name
+ * @param cell binding's cell name within the specifier referenced as "name"
+ * @return the value of the cell inside the named specifier by name
```
Comment (mbolivar-nordic) [inline include/devicetree/clocks.h:219]:
```suggestion
 * @return the value of the cell inside the named specifier
```

File: tests/lib/devicetree/src/main.c [hunk 9]
```diff
@@ -1173,6 +1173,11 @@ static void test_clocks(void)
 			     "TEST_CLOCK"),
 		     "label 0");
 
+	/* DT_CLOCKS_LABEL_BY_NAME */
+	zassert_true(!strcmp(DT_CLOCKS_LABEL_BY_NAME(TEST_TEMP, clk_b),
+			     "TEST_CLOCK"),
+		     "label fixed");
```
Comment (mbolivar-nordic) [inline tests/lib/devicetree/src/main.c:1179]:
```suggestion
		     "label b");
```

File: tests/lib/devicetree/src/main.c [hunk 10]
```diff
@@ -1182,6 +1187,12 @@ static void test_clocks(void)
 		      "clk 2 bits");
 	zassert_equal(DT_CLOCKS_CELL_BY_IDX(TEST_TEMP, bus, 2), 8, "clk 2 bus");
 
+	/* DT_CLOCKS_CELL_BY_NAME */
+	zassert_equal(DT_CLOCKS_CELL_BY_NAME(TEST_TEMP, clk_a, bits), 7,
+		      "clk-a bits");
+	zassert_equal(DT_CLOCKS_CELL_BY_NAME(TEST_TEMP, clk_b, bus), 8,
+		      "clk-a bus");
```
Comment (mbolivar-nordic) [inline tests/lib/devicetree/src/main.c:1194]:
```suggestion
		      "clk-b bus");
```

File: tests/lib/devicetree/src/main.c [hunk 11]
```diff
@@ -1199,6 +1210,11 @@ static void test_clocks(void)
 			     "TEST_CLOCK"),
 		     "label 0");
 
+	/* DT_INST_CLOCKS_LABEL_BY_NAME */
+	zassert_true(!strcmp(DT_INST_CLOCKS_LABEL_BY_NAME(0, clk_b),
+			     "TEST_CLOCK"),
+		     "label fixed");
```
Comment (mbolivar-nordic) [inline tests/lib/devicetree/src/main.c:1216]:
```suggestion
		     "label b");
```

File: tests/lib/devicetree/src/main.c [hunk 12]
```diff
@@ -1208,6 +1224,12 @@ static void test_clocks(void)
 		      "clk 2 bits");
 	zassert_equal(DT_INST_CLOCKS_CELL_BY_IDX(0, bus, 2), 8, "clk 2 bus");
 
+	/* DT_INST_CLOCKS_CELL_BY_NAME */
+	zassert_equal(DT_INST_CLOCKS_CELL_BY_NAME(0, clk_a, bits), 7,
+		      "clk-a bits");
+	zassert_equal(DT_INST_CLOCKS_CELL_BY_NAME(0, clk_b, bus), 8,
+		      "clk-a bus");
```
Comment (mbolivar-nordic) [inline tests/lib/devicetree/src/main.c:1231]:
```suggestion
		      "clk-b bus");
```

File: include/devicetree/clocks.h (hunk 2 above)
Comment (galak) [reply include/devicetree/clocks.h:88]: Was matching what we had in dma.h and adc.h.

File: include/devicetree/clocks.h (hunk 4 above)
Comment (galak) [reply include/devicetree/clocks.h:159]: fixed.

File: include/devicetree/clocks.h (hunk 1 above)
Comment (galak) [reply include/devicetree/clocks.h]: Added a commit in this PR for it.

File: include/devicetree/clocks.h (hunk 2 above)
Comment (mbolivar-nordic) [reply include/devicetree/clocks.h:88]: Ok by me to defer to future cleanup

**Review by mbolivar-nordic — APPROVED** [review]

**Pull request merged**